Fix graph pruning

vepadulano opened the pull request:
### Add abstract class Proxy

The current Proxy class is aimed only at action nodes. It is a wrapper
around those nodes that triggers the execution of the computational graph
and retrieves the value of the action node, when the user issues an action
in the analysis. In order to tackle the graph pruning bug, the current idea
is to wrap all nodes (both transformations and actions) with proxies. With
this in mind, the current `Proxy` class will change name to `ActionProxy`
(for now) and a new abstract class will be created, taking the name of `Proxy`.

The `Proxy` abstract class is initialized with the attribute `node`, that is
the `PyRDF.Node` object linked when the user assigns a new variable to it.

When the user reassigns that same variable to another object (meaning that
he/she has no interest in the previous steps of analysis), the reference count
of the proxy decreases and the `__del__` method is called (automatically). This
abstract method (that will be implemented in both subclasses of the `Proxy`
class) has to flag the node that the current proxy was referring to as
prunable, effectively setting the `PyRDF.Node` attribute `prunable = True`.

Strictly speaking, in this commit no functionality of the project has been
modified. The `PyRDF.Node.prunable` attribute is set to `False` by default and
is only changed by action nodes that are not referenced by the user anymore,
effectively remaining silent. In a later commit this flag will be used to
enhance the `PyRDF.Node.graph_prune()` method.

### Clarify names

Two class attributes previously introduced hadn't a clear meaning.

The attribute `node` in the `Proxy` class has been changed to `proxied_node`,
to better show the link between the proxy object and its wrapped node object.

The attribute `prunable` in the `Node` class was misleading. In fact, one could
think that if a node has `prunable == True`, then that node must be pruned from
the computational graph. This is not the case, since that attribute represents
the fact that a node is wrapped by a proxy which has a direct reference issued by
the user, e.g. via a variable assignment. As a consequence, when a user issues a
command such as `a = df.Filter().Define()`, only the proxy linked to the last
`Define()` node has a user reference, while the `Filter()` proxy doesn't so it
would flag the node as prunable even if it shouldn't be pruned.

For this reason, the attribute `prunable` has been changed to `has_user_references`
to better indicate its true meaning.

### Add TransformationProxy, modify graph pruning logic

A class proxying transformation operations objects has been added, called
`TransformationProxy`, to go alongside with the `ActionProxy` class. Now all
nodes are always wrapped by proxy objects when they are created.

The `TransformationProxy` acts as its corresponding node. In fact, all
functions and attributes calls issued by the user on the proxy get mirrored
back to the node and return methods and attributes of the node itself. As for
the `ActionProxy`, also the `TransformationProxy` flags its wrapped node for
pruning when it gets garbage collected by Python (e.g. when the user is not
assigning a variable to it).

In order to stabilize the results of the graph pruning function, it has been split
in two separate methods. The new method `is_prunable` checks whether a node
can be pruned from the graph. The first condition is that the node has no
children. If this holds, then the attribute `has_user_references` is checked
together with the condition that the node is an action node already computed.
If either of these is `False`, then the method returns `True`. The `graph_prune`
method already present now recurses through the graph, starting from the node
on which is called and evaluating the `is_prunable` method on all of the
children. If a child is prunable, it will not be added to the children list
of the node, otherwise it will.

Other changes:

The main changes of this commit reflect also on some tests, that had to be
changed accordingly. Wherever there was a call to a transformation operation
through a variable, which previously was of class `PyRDF.Node` and now of class
`PyRDF.TransformationProxy`, it has been changed to the attribute `proxied_node`
of the proxy, in order to avoid unmeaningful errors.

Furthermore, it has been noticed that pickling and unpickling a proxy object
through direct call of the functions `pickle.dumps()` / `pickle.loads()`, throws
a `RecursionError` since pickle is not able to reconstruct the link between
proxy and object. This is not considered an error, since our interest is in
pickling the computational graph objects directly without the proxies. In future,
if needed, this could be patched by redefining the `__getstate__` and `__setstate__`
magic methods of the proxy object.

Finally, in `test_dist.py` the function `rangesToTuples` makes use of the
Python `map()` function, which in Python 2 returns a list, while in Python 3
returns an iterator. This behaviour has been fixed by wrapping the returned value
with `list()`.

### Add proxy tests

The newly created proxy classes need to be tested. The new tests check the
following features:

- The abstract class `Proxy` cannot be instantiated by the user
- All subclasses of `Proxy` cannot be instantiated without defining
  the `__del__` abstract method
- The variable assigned to a proxy is of the correct type
- The `proxied_node` attribute of a proxy is of the correct type
- The TransformationProxy correctly reads attributes passed to it,
  recognizes if they are supported operations or other attributes
  and returns the corresponding methods or attributes of the proxied
  node
- The `has_user_references` attribute of a node proxied by a
  `TransformationProxy` object is equal to `True` until the proxy is
  assigned to a variable
- The function `__getstate__` of a `TransformationProxy` object is
  called when the proxy is pickled. This is to check the behaviour
  of pickle, knowing that right now a proxy cannot be unpickled. In
  order to track the calls to the function a decorator has been added
  to the `__getstate__` method of the `TransformationProxy` class.

Commit:
Fix graph pruning

Add abstract class Proxy

The current Proxy class is aimed only at action nodes. It is a wrapper
around those nodes that triggers the execution of the computational graph
and retrieves the value of the action node, when the user issues an action
in the analysis. In order to tackle the graph pruning bug, the current idea
is to wrap all nodes (both transformations and actions) with proxies. With
this in mind, the current `Proxy` class will change name to `ActionProxy`
(for now) and a new abstract class will be created, taking the name of `Proxy`.

The `Proxy` abstract class is initialized with the attribute `node`, that is
the `PyRDF.Node` object linked when the user assigns a new variable to it.

When the user reassigns that same variable to another object (meaning that
he/she has no interest in the previous steps of analysis), the reference count
of the proxy decreases and the `__del__` method is called (automatically). This
abstract method (that will be implemented in both subclasses of the `Proxy`
class) has to flag the node that the current proxy was referring to as
prunable, effectively setting the `PyRDF.Node` attribute `prunable = True`.

Strictly speaking, in this commit no functionality of the project has been
modified. The `PyRDF.Node.prunable` attribute is set to `False` by default and
is only changed by action nodes that are not referenced by the user anymore,
effectively remaining silent. In a later commit this flag will be used to
enhance the `PyRDF.Node.graph_prune()` method.

Clarify names

Two class attributes previously introduced hadn't a clear meaning.

The attribute `node` in the `Proxy` class has been changed to `proxied_node`,
to better show the link between the proxy object and its wrapped node object.

The attribute `prunable` in the `Node` class was misleading. In fact, one could
think that if a node has `prunable == True`, then that node must be pruned from
the computational graph. This is not the case, since that attribute represents
the fact that a node is wrapped by a proxy which has a direct reference issued by
the user, e.g. via a variable assignment. As a consequence, when a user issues a
command such as `a = df.Filter().Define()`, only the proxy linked to the last
`Define()` node has a user reference, while the `Filter()` proxy doesn't so it
would flag the node as prunable even if it shouldn't be pruned.

For this reason, the attribute `prunable` has been changed to `has_user_references`
to better indicate its true meaning.

Add TransformationProxy, modify graph pruning logic

A class proxying transformation operations objects has been added, called
`TransformationProxy`, to go alongside with the `ActionProxy` class. Now all
nodes are always wrapped by proxy objects when they are created.

The `TransformationProxy` acts as its corresponding node. In fact, all
functions and attributes calls issued by the user on the proxy get mirrored
back to the node and return methods and attributes of the node itself. As for
the `ActionProxy`, also the `TransformationProxy` flags its wrapped node for
pruning when it gets garbage collected by Python (e.g. when the user is not
assigning a variable to it).

In order to stabilize the results of the graph pruning function, it has been split
in two separate methods. The new method `is_prunable` checks whether a node
can be pruned from the graph. The first condition is that the node has no
children. If this holds, then the attribute `has_user_references` is checked
together with the condition that the node is an action node already computed.
If either of these is `False`, then the method returns `True`. The `graph_prune`
method already present now recurses through the graph, starting from the node
on which is called and evaluating the `is_prunable` method on all of the
children. If a child is prunable, it will not be added to the children list
of the node, otherwise it will.

Other changes:

The main changes of this commit reflect also on some tests, that had to be
changed accordingly. Wherever there was a call to a transformation operation
through a variable, which previously was of class `PyRDF.Node` and now of class
`PyRDF.TransformationProxy`, it has been changed to the attribute `proxied_node`
of the proxy, in order to avoid unmeaningful errors.

Furthermore, it has been noticed that pickling and unpickling a proxy object
through direct call of the functions `pickle.dumps()` / `pickle.loads()`, throws
a `RecursionError` since pickle is not able to reconstruct the link between
proxy and object. This is not considered an error, since our interest is in
pickling the computational graph objects directly without the proxies. In future,
if needed, this could be patched by redefining the `__getstate__` and `__setstate__`
magic methods of the proxy object.

Finally, in `test_dist.py` the function `rangesToTuples` makes use of the
Python `map()` function, which in Python 2 returns a list, while in Python 3
returns an iterator. This behaviour has been fixed by wrapping the returned value
with `list()`.

Add proxy tests

The newly created proxy classes need to be tested. The new tests check the
following features:

- The abstract class `Proxy` cannot be instantiated by the user
- All subclasses of `Proxy` cannot be instantiated without defining
  the `__del__` abstract method
- The variable assigned to a proxy is of the correct type
- The `proxied_node` attribute of a proxy is of the correct type
- The TransformationProxy correctly reads attributes passed to it,
  recognizes if they are supported operations or other attributes
  and returns the corresponding methods or attributes of the proxied
  node
- The `has_user_references` attribute of a node proxied by a
  `TransformationProxy` object is equal to `True` until the proxy is
  assigned to a variable
- The function `__getstate__` of a `TransformationProxy` object is
  called when the proxy is pickled. This is to check the behaviour
  of pickle, knowing that right now a proxy cannot be unpickled. In
  order to track the calls to the function a decorator has been added
  to the `__getstate__` method of the `TransformationProxy` class.

Fix graph pruning

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -1,8 +1,60 @@
+    return wrapper
+
+
+ABC = ABCMeta('ABC', (object,), {})
```
I would add here a comment to clarify that this is needed to be python2/3 compatible

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -1,8 +1,60 @@
 
 
-class Proxy(object):
+def trackcalls(func):
```
Why do we need this function?

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -1,8 +1,60 @@
+            should wrap.
+        """
+        self.proxied_node = node
+        # self.getstate_called = False
```
This can be removed

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -1,8 +1,60 @@
+        # self.getstate_called = False
+
+    @abstractmethod
+    def __del__(self):
```
Since both Proxy classes implement the same `__del__` method, the implementation could be defined here on the parent class and remove the abstract method.

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -64,14 +106,62 @@ def GetValue(self):
+            return getattr(self.proxied_node, attr)
+
+    @trackcalls
+    def __getstate__(self):
```
I would implement this in a different way. Our main goal is to avoid proxies to be pickled by users, so we could just raise an `Exception` if that happens. 

Now I see what's the purpose of the `trackcall` method, however looks like this allows users to pickle proxies (`p = pickle.dumps(myproxy)` should work without any warning nor error) but the result cannot be pickled back (`pickle.loads(p)`), because the `__setstate__` method is not defined so the `Node.__setstate__` gets called. 

We can just raise an Exception and get rid of `trackcall`, this would also require changes on the test (I will mention it later on this review).

vepadulano replied:
should it raise a `pickle.PickleError()` or should we define a custom exception for that?

etejedor replied:
Why do we need to care about pickling of `Proxy`s at all?

JavierCVilla replied:
Would you suggest to remove every reference to pickle in the `Proxy` class? 

With the current implementation proxies cannot be pickled without a proper implementation of `__setstate__` and `__getstate__` that overrides the Node's behaviour, indeed it leads to a known unstable state and an infinite recursion. This might change after moving the Node's methods to the Proxy class as you proposed above though.

etejedor replied:
My point is that, if `Proxy`s are just an artifact and what contains the real information to unroll the execution is `Node`, we just need to serialize `Node`s.

vepadulano replied:
But we are only serializing `Node`s right now, because they contain all the information. If we instead provide the information on operations and arguments in `Proxy`s I don't know what that would imply as far as the serialization step is concerned

etejedor replied:
I am not proposing to move any information that is needed for the execution to `Proxy`, that has to stay in `Node` of course.
But I think we are diverging from the topic of this particular discussion, which is why we need to serialize `Proxy`s. My point is that we don't and therefore no `__getstate__` is necessary.

JavierCVilla commented on tests/unit/test_node.py:
```diff
@@ -287,7 +287,7 @@ def test_get_state(self):
 
         self.assertDictEqual(node.__getstate__(), node_dict)
-        self.assertDictEqual(n1.__getstate__(), n1_dict)
+        self.assertDictEqual(n1.proxied_node.__getstate__(), n1_dict)
```
I would add a comment explaining that `node` is a `Node` object, while `n1` is indeed a `Proxy` so we need to access to the proxied node before getting the state.

JavierCVilla commented on tests/unit/test_proxy.py:
```diff
@@ -1,9 +1,59 @@
+        with self.assertRaises(TypeError):
+            Proxy()
+
+    def test_subclass_without_method_error(self):
```
If we move `__del__` to `Proxy`, then this test is not needed any more.

JavierCVilla commented on tests/unit/test_node.py:
```diff
@@ -39,13 +39,13 @@ def test_proxy_return(self):
         node = Node(None, None)
         newNode = node.Count()
-        self.assertIsInstance(newNode, Proxy)
+        self.assertIsInstance(newNode, ActionProxy)
```
Even though, there is a similar test for the `Proxy`, I would add also here:

```
self.assertIsInstance(newNode.proxied_node, Node)
```

The reason is that here we are using real operations (`Define`, `Count`, ...) while in the proxy specific tests (`test_proxy.py`) there is a explicit call to `ActionProxy` so adding this second assert we cover both cases.

JavierCVilla commented on tests/unit/test_node.py:
```diff
@@ -39,13 +39,13 @@ def test_proxy_return(self):
         node = Node(None, None)
         newNode = node.Define(1)
-        self.assertIsInstance(newNode, Node)
+        self.assertIsInstance(newNode, TransformationProxy)
```
Same as before but for `TransformationProxy`.

JavierCVilla commented on tests/unit/test_proxy.py:
```diff
@@ -16,15 +66,39 @@ def val(self, arg):
+        proxy = TransformationProxy(node)
+
+        for attr in current_backend.supported_operations:
+            getattr(proxy, attr)
```
`getattr` modifies both the `proxy` object and creates a new `proxy` if `attr` is a supported operation, so both cases should be checked.

Here we need to assign the result of this call to a variable so we can check what has been returned, indeed it should be a new Proxy with a proxied node whose `operation.name` is `attr`. 

The following assert (`self.assertEqual(proxy.proxied_node._cur_attr, attr)` is correct but we need to check also the result of the `getattr` call, otherwise we are just testing half of the logic.

JavierCVilla commented on tests/unit/test_proxy.py:
```diff
@@ -16,15 +66,39 @@ def val(self, arg):
+            self.assertTrue(callable(proxy.proxied_node._call_handler))
+
+        proxy.attribute
+        self.assertEqual(proxy.proxied_node._cur_attr, "attribute")
```
I think this should not work, this would allow users to call non-existing Node attributes.

We have three cases here:
- Proxy.SupportedOperation (`Define`, `Count`, ...): if the attribute is a supported operation, then `_cur_attr` gets modified and holds the name of the incoming operation (previous test)
- Proxy.NodeAttribute: for example `operation` or `children`, we *return* the node's attribute.
- Proxy.NonExistingNodeAttribute: for example `attribute`, raise an `AttributeError` exception

Given the amount of asserts and cases, I would split this tests into different tests, at least one for each case.

JavierCVilla commented on tests/unit/test_proxy.py:
```diff
@@ -16,15 +66,39 @@ def val(self, arg):
+        for attr in current_backend.supported_operations:
+            getattr(proxy, attr)
+            self.assertEqual(proxy.proxied_node._cur_attr, attr)
+            self.assertTrue(callable(proxy.proxied_node._call_handler))
```
What are we testing here? Isn't it always true as long as the proxied node exists?

JavierCVilla commented on tests/unit/test_proxy.py:
```diff
@@ -16,15 +66,39 @@ def val(self, arg):
+        self.assertEqual(proxy.proxied_node._cur_attr, "attribute")
+
+    def test_proxied_node_has_user_references(self):
+        """check that the user reference holds until the proxy lives."""
```
Maybe here we can add the reason: after the garbage collector removes the `proxy`, the `__del__` method modifies this attribute in the node object.

JavierCVilla replied:
Also, try to be consistent with the docstrings, I would start with uppercase all docstrings.

JavierCVilla commented on tests/unit/test_proxy.py:
```diff
@@ -89,3 +162,20 @@ def test_get_value_with_value_not_existing(self):
+
+        node = Node(None, None)
+        proxy = TransformationProxy(node)
+        pickle.dumps(proxy)
```
As mentioned before, this method should raise an Exception, so we do not even allow users to pickle proxies.

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -1,8 +1,60 @@
+    Abstract class for proxies objects. These objects help to keep track of
+    nodes' variable assignment. That is, when a node is no longer assigned
+    to a variable by the user, the role of the proxy is to flag that node as
+    `prunable`. There are two main classes for proxies, depending on the
```
`prunable` was the previous name for the attribute, but now we use `has_user_references`, so this might be misleading for users.

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -1,8 +1,60 @@
+    nodes' variable assignment. That is, when a node is no longer assigned
+    to a variable by the user, the role of the proxy is to flag that node as
+    `prunable`. There are two main classes for proxies, depending on the
+    operation type of the node they are wrapping:
```
I would not mention the possible subclasses here. They are already documented down on the same file.

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -1,8 +1,60 @@
+        - ActionProxy: a proxy wrapping a node that holds an action operation.
+        - TransformationProxy: a proxy wrapping a node that holds a
+        transformation operation.
+    """
```
We also need to advice users that Proxies cannot be _pickled_.

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -1,8 +1,60 @@
+
+    def __init__(self, node):
+        """
+        Creates a new `Proxy` object for a
```
This could be put in one single line.

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -1,8 +1,60 @@
+        Parameters
+        ----------
+        proxied_node : PyRDF.Node
+            The node that the current Proxy
```
This could be put in one single line.

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -1,8 +1,60 @@
+    @abstractmethod
+    def __del__(self):
+        """
+        Proxy has to flag a node as prunable when the user changes
```
This description can be misleading since there are several conditions to be pruned.

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -15,26 +67,12 @@ class Proxy(object):
 
-    action_node
+    node
         The action node that the current Proxy
```
I would put this in one line if it fits.

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -50,6 +88,10 @@ def __getattr__(self, attr):
         return self._call_handler
 
+    def __del__(self):
+        """Deletes current Proxy and flags the wrapped Node for pruning"""
```
`__del__` does not really _delete_ the proxy, but defines the actions to be done right before getting deleted. Also, `for pruning` can be misleading, since there are more conditions. You could put something like:

_Flag the wrapper node when there is no more user references pointing to it, which is one of the conditions to be pruned from the internal graph._

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -64,14 +106,62 @@ def GetValue(self):
+    user references anymore.
+    """
+
+    def __del__(self):
```
To be moved to the parent class.

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -64,14 +106,62 @@ def GetValue(self):
+        Intercepts calls on operation or attributes belonging to the proxied
+        node.
+
+        Returns either:
```
This should be documented like [here](https://github.com/JavierCVilla/PyRDF/blob/master/PyRDF/RDataFrame.py#L127-L134).

JavierCVilla reviewed (changes requested):
Ciao @vepadulano , thank you very much for this PR! You got the idea super fast.

Please find below some comments.

Commit:
Add comments, tests for proxies

The ABC declaration was a bit misleading. Now a comment has been added to explain that
Python 2 does not implement the ABC class directly, but only the ABCMeta class.

Some tests have been added to check that TransformationProxy instances return the correct
node attributes when they are called. For instance, if a transformation operation is called
on a proxy, it will create a new proxy to a node holding that operation. On the other hand,
if a node class attribute is called on a proxy, this should return the proxied node attribute.
Finally, if a non-defined attribute is called on a proxy, this should raise an `AttributeError`.

The pickling test has been modified. While before it tested that the `__getstate__` method of
the TransformationProxy class was called when issuing a pickle.dumps() on the proxy, now it
tests that pickling a proxy raises a `pickle.PickleError`, since we are not providing that
functionality at the moment.

etejedor commented on PyRDF/Proxy.py:
```diff
@@ -62,16 +84,56 @@ def GetValue(self):
+    user references anymore.
+    """
+
+    def __getattr__(self, attr):
```
Since we are not getting attributes on `Node`s anymore (because now we always interact with a `Proxy` from the user application), the logic of the `__getattr__` in `Node` should be moved to `Proxy`. And the same thing for `_call_handler`.
It will require some code refactoring to understand what is common to both subclasses of `Proxy` and what needs to be done only in a specific subclass.

vepadulano replied:
I believe that such a refactoring would defeat the purpose of a proxy. The proxy should just mirror the methods and attributes of the node and maybe add little modifications (like the `has_user_references` attribute). But then, if all operation got directly managed and returned by proxies, what would be the purpose of the node? Furthermore, to keep all the information about the computational graph (callables included) we would also need to send information about proxies to the distributed executors (possibly via pickling) and right now we are not allowing that

etejedor replied:
It depends on what functionality is outsourced to the `Proxy`. See how in `_call_handler` of `Node` we are going back to creating a `Proxy`:
https://github.com/JavierCVilla/PyRDF/blob/fbf79aec546dd816c901cbcd397f677b6ed80f88/PyRDF/Node.py#L167
This is an indicator that it should be the proxy itself that creates the node and stores it in its belly.
See also this:
https://github.com/JavierCVilla/PyRDF/blob/fbf79aec546dd816c901cbcd397f677b6ed80f88/PyRDF/Node.py#L143
such check is repeated here:
https://github.com/JavierCVilla/PyRDF/blob/fbf79aec546dd816c901cbcd397f677b6ed80f88/PyRDF/Proxy.py#L127
Having the `__getattr__` and `_call_handler` functionality split in two classes is not ideal, I would just have it in the class where you get the attribute and where you actually do the call.

In the end, `Node` can just be a class that is serialized to the backend and contains all the information needed for the execution. `Proxy` can take care of the checks and creation of the corresponding `Node`.

Commit:
switch operation logic to proxy

The current implementation relies on the nodes to handle all attribute
and method calls, then create their own proxies when needed. The proxies
then have to return again the correct attributes and methods of their
wrapped nodes. This makes for logic and code-wise duplication that is
not needed.

In order to tackle this situation, a decision was made to leave the
creation of the computational graph (nodes and operations), together
with the subsequent access to attributes and methods, to proxies. Then
the nodes would only be responsible for storing the information needed
by the backend to execute the analysis of the user.

For this reason, the `__getattr__` and `_call_handler` methods of the
`Node` class have been transferred to the `Proxy` class. With this new
logic, whenever a user calls an attribute or a method on a node, the
proxy will get the right attribute or method of the node. If the user
called for an operation the proxy would then create a new node with that
specific operation, thus adding new pieces to the graph.

The only exception to this scheme was the head node. For all intents
and purposes, this must be an `RDataFrame` object, a subclass
of the `Node` class. In order to avoid unnecessary links between this
subclass and the `Proxy` class, the new `RDataFrame` class is just
a means to kickstart the creation of the head node of the computational
graph. In fact, its `__new__` method creates an instance of the
`HeadNode` class (which is just the old `RDataFrame` with a new name)
with the arguments passed by the user, then wraps it with a proxy and
finally returns the proxy itself.

Note that almost all the code could be implemented in the parent
`Proxy` class, but logically there is a difference between action and
transformation nodes and now there is also the head node, that is not
an operation at all.

Another reason for the current implementation of the `Proxy` class is
that through the `__getattr__` method there is a means to check what
operation the user is calling on the data and only then return the right
proxy. In conclusion, there is surely the need to change the name of
`TransformationProxy` class and possibly also to decide wether to keep
many subclasses of `Proxy` at all.

Commit:
Add documentation, change tests

In order to reflect the change of logic of the last commit, some
docstrings have been added to `RDataFrame` class and to the renewed
proxy family of classes.

Furthermore, many tests have been modified in order to account for the
fact that now all attributes and methods of nodes are accessed through
their proxies.

A minor issue has been noticed, that in Python 3 abstract classes that
do not implement abstract methods can be instantiated anyway, so for the
moment there is no need to test that the `Proxy` class cannot be
instantiated. This will be addressed in later commits.

Commit:
Clear division of tasks in proxy subclasses

The tasks that the family of proxy classes should manage now are more clearly assigned
to either the `TransformationProxy` or `ActionProxy` subclasses.

All nodes that make changes to data (The head node that collects new data, the
transformations that modify their structure) are proxied by `TransformationProxy`s.
Only nodes that hold action operations are proxied by `ActionProxy`s.

The first subclass handles checks on the attribute called on the node. If it is a node
attribute, the proxy will return its value, or if it is an operation name it will return
the appropriate proxy for that operation, after creating the node and the operation
objects linked to it.

The second subclass keeps its original behaviour, attribute calls return the value
of the action, or trigger the execution of the graph if it has not been done yet.

The `Proxy` parent class is now an abstract class with abstract method `__getattr__`.

JavierCVilla commented on PyRDF/Node.py:
```diff
@@ -85,6 +73,7 @@ def __init__(self, get_head, operation):
         self._cur_attr = ""  # Name of the new incoming operation
         self.value = None
         self.pyroot_node = None
+        self.has_user_references = True  # Flag for pruning
```
This comment and the one in `_cut_attr` can be removed, it is already documented on the class.

vepadulano replied:
ok

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -1,54 +1,90 @@
+        # return the result of the action
+        try:
+            if self.proxied_node.operation.is_action():
+                self._cur_op_call = attr  # Stores the name of operation call
```
The `self._cur_op_call` attribute needs to be documented.

vepadulano replied:
this is outdated, it refers to the commit where I removed the proxy subclasses. Now is reverted back to `_cur_attr` attribute for the `ActionProxy` class. Still don't like it though, would prefer to call it `cur_action_attr` or `cur_action_call` because it's the string we pass to the `GetValue` method.

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -1,54 +1,90 @@
+
+    def _create_new_op(self, *args, **kwargs):
+        """
+        Handles an operation call to the current node and eturns the new node
```
Typo in `eturns`

vepadulano replied:
yes, missed it!

JavierCVilla commented on PyRDF/Node.py:
```diff
@@ -29,6 +26,10 @@ class Node(object):
         children nodes connected to the current node.
 
+    _cur_attr
+        The name of the operation of the last child node among
```
I would put the comment used below which is bit clearer: `Name of the new incoming operation`

vepadulano replied:
To me it's not that clear, still that can be changed easily as you wish

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -1,40 +1,66 @@
+
+    Attributes
+    ----------
+    backend
```
Is `backend` a class member?

vepadulano replied:
No it isn't! Don't even know why that's there, it's inherited from the documentation in the master branch

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -1,40 +1,66 @@
+        A class member to store a backend object based on the configuration
+        set by the user.
 
+    proxied_node
```
This attribute is already documented in the parent class so it can be removed from the subclass.

vepadulano replied:
ok

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -1,40 +1,66 @@
+    @abstractmethod
+    def __getattr__(self, attr):
+        """
+        Proxies have to declare a way they handle attributes called on them.
```
I would put _...the way they handle calls to attributes and methods of the proxied nodes_.

JavierCVilla replied:
Or even better, I would use the description used in the `TransformationProxy` implementation: 

```
Intercepts calls to attributes and methods of the proxied node and returns the appropriate object(s).
```

vepadulano replied:
changed the docstring to
```
Proxies have to declare the way they intercept calls to attributes
        and methods of the proxied node.
```

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -62,16 +88,77 @@ def GetValue(self):
+
+class TransformationProxy(Proxy):
+    """
+    A proxy object to an instantiated node. Used as a controller of the user
```
Here I would be more specific about what `TransformationProxy` does different with respect to `Proxy`. For example: 

```
A proxy object to non-action nodes. It implements access to the 
proxied_node's attributes and the creation of new graph nodes.
```

vepadulano replied:
done

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -62,16 +88,77 @@ def GetValue(self):
+        # import re
+        # if re.search("^__[a-z]+__$", attr):
+        #     # Raise an AttributeError for all dunder method calls
+        #     raise AttributeError("Such an attribute is not set ! ")
```
Can this comment be already removed?

vepadulano replied:
yes actually, we want transformation proxies to return all node attributes when requested.

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -62,16 +88,77 @@ def GetValue(self):
+        # operation and node creation
+        if attr in current_backend.supported_operations:
+            self.proxied_node._cur_attr = attr  # Stores new operation name
+            return self._call_handler
```
I would use the name `_create_new_op` rather than `_call_handler` as you did in a previous commit, I think it's clearer. In general, I think the names introduced in the commit that you reverted were quite well defined.

vepadulano replied:
ok thanks, changed them

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -62,16 +88,77 @@ def GetValue(self):
+            try:
+                return getattr(self.proxied_node, attr)
+            except AttributeError as e:
+                raise e
```
What's the error message that gets printed if we reach this point? Is it worth to raise a custom message?

vepadulano replied:
it prints that the attribute is not defined. If the `except` statement is left blank, it would go against PEP8 style rules. In fact, flake8 gives the error `E722 do not use bare 'except'`. If you think we shouldn't raise the error message, we could just print something to let the user know that's not a defined attribute, but I'm not sure that'd be worthwhile

vepadulano replied:
Forcing the error gives such a message:
```
>>> a.randomattr
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "PyRDF/Proxy.py", line 128, in __getattr__
    raise e
AttributeError: 'Node' object has no attribute 'randomattr'

```

JavierCVilla commented on PyRDF/Proxy.py:
```diff
@@ -62,16 +88,77 @@ def GetValue(self):
+            except AttributeError as e:
+                raise e
+
+    def _call_handler(self, *args, **kwargs):
```
As mentioned before I would rename this method.

vepadulano replied:
renamed to `_create_new_op`

JavierCVilla commented on PyRDF/RDataFrame.py:
```diff
@@ -1,9 +1,37 @@
+    The user call to create an RDataFrame will issue a call to the `__new__`
+    method. In turn, it will create an `HeadNode` object with the arguments
+    used to create the RDataFrame. The head node will then be wrapped with a
+    `TransformationProxy` object (for now) and then the `__new__` method will
```
I would remove `(for now)`.

vepadulano replied:
done

JavierCVilla commented on PyRDF/RDataFrame.py:
```diff
@@ -55,7 +81,7 @@ def __init__(self, *args):
                 args[i] = self._get_vector_from_list(args[i])
 
         try:
-            ROOT.ROOT.RDataFrame(*args)  # Check if the args are correct
```
I would prefer to keep this comment, since it might not be clear for new users why we do it at this point if we claim to be lazy.

vepadulano replied:
ok

JavierCVilla commented on PyRDF/backend/Utils.py:
```diff
@@ -17,6 +17,7 @@ def declare_headers(cls, includes):
         """
         for header in includes:
             include_code = "#include \"{}\"\n".format(header)
+            print(include_code)
```
Is this print needed?

vepadulano replied:
nope, missed it!

JavierCVilla commented on tests/unit/test_callable_generator.py:
```diff
@@ -40,8 +41,7 @@ def test_mapper_from_graph(self):
         # Head node
-        node = Node(None, None)
-
+        node = TransformationProxy(Node(None, None))
```
Since `node` is now a `Proxy`, in the line 53 of this same file, shouldn't we have something like this (even though it work now)? We want to pass the node instead of the proxy:

```
# Generate and execute the mapper
generator = CallableGenerator(node.proxied_node)
```

vepadulano replied:
So the `CallableGenerator` accesses attributes and methods of the node(s), such as `graph_prune` or `operation`. Since (transformation) proxies return all attributes and methods of their proxied node, the tests now works regardless of what object we pass. 

Being said that, I think that as it is now it double checks both the fact that the proxy returns the correct attributes and methods of its node and that the node behaves correctly. Nevertheless, this makes it less clear that what the `CallableGenerator` class handles are nodes, not proxies.

vepadulano replied:
In the end, some more clarity can play to our advantage. I will got with the call to the proxied node

JavierCVilla commented on tests/unit/test_callable_generator.py:
```diff
@@ -71,7 +71,7 @@ def test_mapper_with_pruning(self):
 
         # Head node
-        node = Node(None, None)
+        node = TransformationProxy(Node(None, None))
```
Same as before, should we pass `node.proxied_node` to `CallableGenerator`?

vepadulano replied:
Yes, see my previous comment

JavierCVilla commented on tests/unit/test_node.py:
```diff
@@ -274,29 +275,31 @@ def test_get_state(self):
             'children': []
         }
-
+        # node is of class Node, its state dictionary can be directly accessed.
```
This does not hold true, `node` is of class `Proxy`, so the state has to be accessed through the `proxied_node`

vepadulano replied:
Very similar to the situation of the callable generator above. `node` is of class `TransformationProxy`, but the dictionary of the proxied node can actually be directly accesed since the proxy returns all attributes and methods of the node. For the sake of clarity I will put it explicitly

JavierCVilla commented on tests/unit/test_node.py:
```diff
@@ -274,29 +275,31 @@ def test_get_state(self):
         """
-        node = Node(None, None)  # Head node
-        n1 = Node(None, None)  # First child node
+        node = TransformationProxy(Node(None, None))  # Head node
```
In lines 313 and 314, we need to call the `__set_state__` method of the `proxied_node`.

vepadulano replied:
done

JavierCVilla commented on tests/unit/test_proxy.py:
```diff
@@ -1,61 +1,162 @@
 import unittest
 import PyRDF
 
+# As of Python 3, abstract classes that do not implement
```
Using `ABCMeta` as a parent class of `Proxy` and the decorator `@abstractmethod` this test should work, can you check with the current implementation?

vepadulano replied:
Yes, now the test works properly

JavierCVilla commented on tests/unit/test_proxy.py:
```diff
@@ -1,61 +1,162 @@
         self.assertEqual(proxy._cur_attr, "attr")
         self.assertTrue(callable(func))
 
+    def test_supported_transformation(self):
```
Very good tests :+1:

vepadulano replied:
thanks!

JavierCVilla reviewed (changes requested):
Very nice work @vepadulano, this is getting closer to be merged. Some comments below to polish the code.

JavierCVilla reviewed (approved):
Very well done @vepadulano , looks good to me.
